### DISCUSSION TITLE
Custom error messages for `ValidationElement`

### DISCUSSION THREAD

**Discussion by adosikas** [zauberzeug/nicegui#2150] [Ideas / Feature Requests]
Validation functions should be able to return a string describing the problem, allowing more complex functions to be implemented:

```py
def _validate_json(value: str) -> bool|str:
  try:
    j = json.loads(value)
    j["test_key"]
  except json.decoder.JSONDecodeError as jde:
    return f"Invalid JSON: {jde}"
  except KeyError as ke:
    return f"Missing key: {ke}"
  return True  # or maybe None?
```

The problem is that non-empty strings evalulate as "True", which may break some code that hooks into validations directly (ie https://github.com/zauberzeug/nicegui/discussions/1791#discussioncomment-7270147).

So maybe make this separate from the existing `validation` dict. As the message is no longer needed, the new parameter could just be a list/tuple, which would also allow explicit ordering of validation functions without relying on dict insertion order.

Seems straightforward to implement either way, so I may take that up.

**Comment by falkoschindler** [comment]
Sorry, I don't understand the problem you're trying to solve. Can you share a code example demonstrating the problem and/or the solution? What is `_validate_json` supposed to do? How does it relate to a `ValidationElement`? Thanks! 🙂

**Comment by adosikas** [comment]
Right now, a validation function (ie for `ui.input(... validation={"error": bool_returning_func})`) can only report a single error (the dict key).

For my usecase (validating some JSON string), I would need a lot of validation functions for all the different errors I want to show, all of which would do duplicate work, ie first one tries parsing the string into a JSON, then the second one would parse the JSON and check if key A exists, the third one would need to parse the JSON and check if key B exists, etc.

The current system works for simple "you violated one constraint" errors, but nothing more advanced, which would be possible when the error could be adjusted. Of course I could just make a subclass, but I figured other people may find this useful too. 

```py
# current
def _is_valid_json(s: str) -> bool:
  try:
    json.loads(s)
    return True
  except Exception as e:
    return False
ui.input("Enter JSON", validation={"Invalid JSON": _is_valid_json})  # this error isn't very useful to fix the input
# proposal
def _validate_json(value: str) -> bool|str:
  try:
    j = json.loads(value)
    j["test_key"]
  except json.decoder.JSONDecodeError as jde:
    return f"Invalid JSON: {jde}"
  except KeyError as ke:
    return f"Missing key: {ke}"
  return True  # or maybe None?

# ValidationElement checks if the parameter is a dict or list and uses the new behavior only for lists
# (to be fully backwards compatible)
ui.input("Enter JSON", validation=[_validate_json])
```

**Reply by falkoschindler** [reply]
Oh, I see. This use case is actually part of the ongoing PR #2133. We're just discussing the implementation and which code change is really necessary and what could be simplified.

**Reply by adosikas** [reply]
Oh. Didn't see that.
